src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/reader/MasterEventMetricsSnapshot.java

```diff
@@ -0,0 +1,311 @@
+        } };
+
+        ArrayList<Field<?>> groupByFields = new ArrayList<Field<?>>() { {
+            this.add(DSL.field(DSL.name(AllMetrics.Master_Metric_Dimensions.MASTER_TASK_INSERT_ORDER.toString()), String.class));
+        } };
+
+        return create.select(fields).from(fetchRunTimeHelper())
+                .groupBy(groupByFields)
+                .fetch();
+    }
+
+    public SelectHavingStep<Record> fetchRunTimeHelper() {
+
+        List<SelectField<?>> fields = new ArrayList<SelectField<?>>() { {
+            this.add(DSL.field(DSL.name(AllMetrics.Master_Metric_Dimensions.MASTER_TASK_INSERT_ORDER.toString()), String.class));
+            this.add(DSL.field(DSL.name(AllMetrics.Master_Metric_Dimensions.MASTER_TASK_PRIORITY.toString()), String.class));
+            this.add(DSL.field(DSL.name(AllMetrics.Master_Metric_Dimensions.MASTER_TASK_TYPE.toString()), String.class));
+            this.add(DSL.field(DSL.name(AllMetrics.Master_Metric_Dimensions.MASTER_TASK_METADATA.toString()), String.class));
+            this.add(DSL.field(DSL.name(AllMetrics.Master_Metric_Dimensions.MASTER_TASK_QUEUE_TIME.toString()), String.class));
+            this.add(DSL.field(Fields.ET.toString()).minus(DSL.field(Fields.ST.toString())).
+                    as(DSL.name(AllMetrics.Master_Metric_Dimensions.MASTER_TASK_RUN_TIME.toString())));
+        } };
+
+        return create.select(fields).from(groupByInsertOrderWithoutNull())
+                .where(DSL.field(Fields.ET.toString()).isNotNull().and(
+                        DSL.field(Fields.ST.toString()).isNotNull()));
+    }
+
+
+    /**
+     * Return one row per master task event. Group by the InsertOrder.
+     * For a master task without a finish event, we will use the current window end time
+     *
+     * CurrentWindowEndTime: 1535065341025
+     * Actual Table
+     * |tid  |insertOrder|taskType    |priority|queueTime|metadata|           st|           et|
+     * +-----+-----------+------------+--------+---------+--------+-------------+-------------+
+     * |111  |1          |create-index|urgent  |3        |{string}|1535065340625|       {null}|
+     * |111  |2          |create-index|urgent  |12       |{string}|1535065340825|       {null}|
+     * |111  |1          |      {null}|  {null}|   {null}|  {null}|       {null}|1535065340725|
+     *
+     * Returned:
+     *
+     * |tid  |insertOrder|taskType    |priority|queueTime|metadata|           st|           et|
+     * +-----+-----------+------------+--------+---------+--------+-------------+-------------+
+     * |111  |1          |create-index|urgent  |3        |{string}|1535065340625|1535065340725|
+     * |111  |2          |create-index|urgent  |12       |{string}|1535065340825|1535065341025|
+     *
+     * @return aggregated master task
+     */
+    public SelectHavingStep<Record> groupByInsertOrderWithoutNull() {
```

Comment:
Can we rename this function to indicate that this will pick the window endtime if endtime does not exist? Maybe something like selectRuntimeTable()?

Reply:
Somehow if I removed groupBy here, we will produce all null rows, and adding where clause does not get rid of null rows, I will just keep this for now.